### PR TITLE
Code inspection changes with Tim/Abhipsa

### PR DESCRIPTION
**Pull request by drwill-ms**

### CODE REVIEW

File: iothub/device/src/Transport/Amqp/AmqpUnit.cs [hunk 1]
```diff
@@ -7,12 +7,23 @@
...
 using Microsoft.Azure.Devices.Client.Transport.Amqp;
 
 namespace Microsoft.Azure.Devices.Client.Transport.AmqpIot
 {
     internal class AmqpUnit : IDisposable
     {
+        // AMQP supports a few different ways to acknowledge a message. Below is the mapping
+        // from the IoT hub level concept of Complete/Abandon/Reject to the AMQP level concept
+        // of Accepted/Released/Rejected.
+        private static readonly IReadOnlyDictionary<MessageAcknowledgement, Outcome> s_ackMap = new Dictionary<MessageAcknowledgement, Outcome>
```
Comment (drwill-ms) [inline iothub/device/src/Transport/Amqp/AmqpUnit.cs:20]: Moved from if/elseif/else to a map.

File: iothub/device/src/Transport/Amqp/AmqpUnit.cs [hunk 2]
```diff
@@ -196,23 +206,12 @@ internal async Task<AmqpIotSession> EnsureSessionIsOpenAsync(CancellationToken c
             return _amqpIotSession;
         }
 
-        public async Task CloseAsync(CancellationToken cancellationToken)
+        internal async Task CloseAsync(CancellationToken cancellationToken)
         {
             if (Logging.IsEnabled)
                 Logging.Enter(this, nameof(CloseAsync));
 
-            try
```
Comment (drwill-ms) [inline iothub/device/src/Transport/Amqp/AmqpUnit.cs:204]: We don't put try/catch/throw around semaphore waits in other places, so removing it here too, for consistency.

File: iothub/device/src/Transport/Amqp/AmqpUnit.cs [hunk 3]
```diff
@@ -266,84 +266,13 @@ private bool isPooled()
 
         #region Message
 
-        private async Task EnsureMessageReceivingLinkIsOpenAsync(CancellationToken cancellationToken, bool enableCallback = false)
```
Comment (drwill-ms) [inline iothub/device/src/Transport/Amqp/AmqpUnit.cs:269]: private methods moved below internals

File: iothub/device/src/Transport/Amqp/AmqpUnit.cs [hunk 4]
```diff
@@ -374,101 +304,68 @@ public async Task<AmqpIotOutcome> SendMessageAsync(OutgoingMessage message, Canc
...
+        internal async Task EnableReceiveMessageAsync(CancellationToken cancellationToken)
         {
-            if (_closed)
+            if (Logging.IsEnabled)
+                Logging.Enter(this, nameof(EnableReceiveMessageAsync));
+
+            if (_isClosed)
             {
                 throw new IotHubClientException("Device is now offline.", false);
             }
```
Comment (drwill-ms) [inline iothub/device/src/Transport/Amqp/AmqpUnit.cs:315]: Logging before error handling.

File: iothub/device/src/Transport/Amqp/AmqpUnit.cs [hunk 5]
```diff
@@ -981,6 +869,51 @@ await Task
...
+                        break;
+
+                    case AmqpTwinMessageType.Put:
+                        amqpIotOutcome = await _twinSendingLink
+                            .SubscribeToDesiredPropertiesAsync(correlationId, cancellationToken)
+                            .ConfigureAwait(false);
+                        break;
+                }
+
+                amqpIotOutcome?.ThrowIfNotAccepted();
```
Comment (drwill-ms) [inline iothub/device/src/Transport/Amqp/AmqpUnit.cs:908]: This was repeated in each case, so centralized to here after the switch block.

File: iothub/device/src/Transport/AmqpIot/AmqpIotMessageConverter.cs [hunk 6]
```diff
@@ -39,17 +34,19 @@ public static IncomingMessage AmqpMessageToIncomingMessage(AmqpMessage amqpMessa
             Argument.AssertNotNull(amqpMessage, nameof(amqpMessage));
 
             using var ms = new MemoryStream();
-            amqpMessage.BodyStream.CopyTo(ms);
-
-            var message = new IncomingMessage(ms.ToArray())
+            using (amqpMessage)
```
Comment (drwill-ms) [inline iothub/device/src/Transport/AmqpIot/AmqpIotMessageConverter.cs:37]: Changed to use a more elegant dispoing scheme.

File: iothub/device/src/Transport/AmqpIot/AmqpIotOutcome.cs [hunk 7]
```diff
@@ -22,28 +21,5 @@ public void ThrowIfNotAccepted()
                 throw AmqpIotErrorAdapter.GetExceptionFromOutcome(_outcome);
             }
         }
-
-        public void ThrowIfError()
```
Comment (drwill-ms) [inline iothub/device/src/Transport/AmqpIot/AmqpIotOutcome.cs:26]: Unused

File: iothub/device/src/Transport/AmqpIot/AmqpIotReceivingLink.cs [hunk 8]
```diff
@@ -60,37 +61,23 @@ internal void SafeClose()
...
+
+            Debug.Assert(disposeOutcome is Accepted, "IoT hub rejected the ack, which we don't expect and if we find it does, we should handle it.");
 
             if (Logging.IsEnabled)
                 Logging.Exit(this, outcome, nameof(DisposeMessageAsync));
-
-            return new AmqpIotOutcome(disposeOutcome);
-        }
-
-        private static ArraySegment<byte> ConvertToDeliveryTag(string lockToken)
```
Comment (drwill-ms) [inline iothub/device/src/Transport/AmqpIot/AmqpIotReceivingLink.cs:81]: Unused

File: iothub/device/src/Transport/AmqpIot/AmqpIotResultAdapter.cs [hunk 9]
```diff
@@ -1,29 +0,0 @@
...
-        Rejected
-    }
-
-    internal static class AmqpIotResultAdapter
-    {
-        internal static Outcome GetResult(AmqpIotDisposeActions amqpIotConstants)
-        {
-            return amqpIotConstants switch
-            {
-                AmqpIotDisposeActions.Accepted => new Accepted(),
```
Comment (drwill-ms) [inline iothub/device/src/Transport/AmqpIot/AmqpIotResultAdapter.cs:22]: Logic moved to a map

**Review by timtay-microsoft — APPROVED** [review]

**Pull request merged**